### PR TITLE
Remove T parameter from Kernel

### PR DESCRIPTION
**Pull request by willtebbutt**
As discussed in #18

### CODE REVIEW

File: src/kernels/matern.jl [hunk 1]
```diff
@@ -44,15 +35,11 @@ The matern 3/2 kernel is an isotropic Mercer kernel given by the formula:
...
-struct Matern32Kernel{T,Tr} <: Kernel{T,Tr}
+struct Matern32Kernel{Tr} <: Kernel{Tr}
     transform::Tr
-
-    function Matern32Kernel{T,Tr}(transform::Tr) where {T,Tr<:Transform}
-        return new{T,Tr}(transform)
-    end
 end
 
-@inline kappa(κ::Matern32Kernel, d::T) where {T<:Real} = (1+sqrt(3)*d)*exp(-sqrt(3)*d)
```
Comment (theogf) [inline src/kernels/matern.jl]: Could you leave `T` here? I will need to replace 1 by `one(T)`

File: src/kernels/matern.jl [hunk 2]
```diff
@@ -63,14 +50,10 @@ The matern 5/2 kernel is an isotropic Mercer kernel given by the formula:
...
-struct Matern52Kernel{T,Tr} <: Kernel{T,Tr}
+struct Matern52Kernel{Tr} <: Kernel{Tr}
     transform::Tr
-
-    function Matern52Kernel{T,Tr}(transform::Tr) where {T,Tr<:Transform}
-        return new{T,Tr}(transform)
-    end
 end
 
-@inline kappa(κ::Matern52Kernel, d::Real) where {T} = (1+sqrt(5)*d+5*d^2/3)*exp(-sqrt(5)*d)
```
Comment (theogf) [inline src/kernels/matern.jl]: Same here (although it's inconsistent :sweat_smile: )

File: src/kernels/constant.jl [hunk 3]
```diff
@@ -3,7 +3,7 @@ ZeroKernel([tr=IdentityTransform()])
 
 Create a kernel always returning zero
 """
-struct ZeroKernel{T,Tr} <: Kernel{T,Tr}
+struct ZeroKernel{T,Tr} <: Kernel{Tr}
```
[the remark this answers is not in the document]
Comment (willtebbutt) [reply src/kernels/constant.jl]: Ideally I would, but the output type isn't determined by the inputs so I think we still need to know what type to output.

File: src/kernels/constant.jl [hunk 4]
```diff
@@ -27,7 +27,7 @@ metric(::ZeroKernel) = Delta()
 ```
 Kernel function working as an equivalent to add white noise.
 """
-struct WhiteKernel{T,Tr} <: Kernel{T,Tr}
+struct WhiteKernel{T,Tr} <: Kernel{Tr}
```
[the remark this answers is not in the document]
Comment (willtebbutt) [reply src/kernels/constant.jl]: Same as for `ZeroKernel`

File: src/kernels/matern.jl (hunk 1 above)
Comment (willtebbutt) [reply src/kernels/matern.jl]: I'm not sure that you actually need to make this change -- the promotion mechanism / constant prop seems to sort stuff out Consider:
```julia
kappa(d::Real) = (1+sqrt(3)*d)*exp(-sqrt(3)*d)
kappa_new(d::Real) = (one(d)+sqrt(3)*d)*exp(-sqrt(3)*d)
```
The native code produced is identical:
```julia
julia> @code_native kappa(0.5)
	.section	__TEXT,__text,regular,pure_instructions
; ┌ @ REPL[2]:1 within `kappa'
	pushq	%rax
	movabsq	$4823801376, %rax       ## imm = 0x11F855E20
; │┌ @ float.jl:405 within `*'
	vmulsd	(%rax), %xmm0, %xmm1
	movabsq	$4823801384, %rax       ## imm = 0x11F855E28
; │└
; │┌ @ promotion.jl:311 within `+' @ float.jl:401
	vaddsd	(%rax), %xmm1, %xmm1
	vmovsd	%xmm1, (%rsp)
	movabsq	$4823801392, %rax       ## imm = 0x11F855E30
; │└
; │┌ @ float.jl:405 within `*'
	vmulsd	(%rax), %xmm0, %xmm0
; │└
	movabsq	$"reinterpret;", %rax
	callq	*%rax
; │┌ @ float.jl:405 within `*'
	vmulsd	(%rsp), %xmm0, %xmm0
; │└
	popq	%rax
	retq
	nopw	%cs:(%rax,%rax)
; └

julia> @code_native kappa_new(0.5)
	.section	__TEXT,__text,regular,pure_instructions
; ┌ @ REPL[3]:1 within `kappa_new'
	pushq	%rax
	movabsq	$4823816416, %rax       ## imm = 0x11F8598E0
; │┌ @ float.jl:405 within `*'
	vmulsd	(%rax), %xmm0, %xmm1
	movabsq	$4823816424, %rax       ## imm = 0x11F8598E8
; │└
; │┌ @ float.jl:401 within `+'
	vaddsd	(%rax), %xmm1, %xmm1
	vmovsd	%xmm1, (%rsp)
	movabsq	$4823816432, %rax       ## imm = 0x11F8598F0
; │└
; │┌ @ float.jl:405 within `*'
	vmulsd	(%rax), %xmm0, %xmm0
; │└
	movabsq	$"reinterpret;", %rax
	callq	*%rax
; │┌ @ float.jl:405 within `*'
	vmulsd	(%rsp), %xmm0, %xmm0
; │└
	popq	%rax
	retq
	nopw	%cs:(%rax,%rax)
; └
```

Seems to be fine?

Comment (theogf) [reply src/kernels/matern.jl]: True! I guess I am always underestimating Julia compiling mechanisms...

File: src/kernels/constant.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (theogf) [reply src/kernels/constant.jl]: No it is!
https://github.com/theogf/KernelFunctions.jl/blob/faa89cb37e0cd3018b9cf135d5d328e9023d293d/src/kernels/constant.jl#L18

[the remark this answers is not in the document]
Comment (willtebbutt) [reply src/kernels/constant.jl]: Oh, okay. Not sure how `d` knows what type to be, but I'm happy to roll with it :)

File: src/kernels/matern.jl [hunk 5]
```diff
@@ -41,6 +41,9 @@ end
 
 @inline kappa(κ::Matern32Kernel, d::Real) = (1+sqrt(3)*d)*exp(-sqrt(3)*d)
 
+kappa(d::Real) = (1+sqrt(3)*d)*exp(-sqrt(3)*d)
+kappa_new(d::Real) = (one(d)+sqrt(3)*d)*exp(-sqrt(3)*d)
```
Comment (theogf) [inline src/kernels/matern.jl]: You left your kappa test there ^^

Comment (willtebbutt) [reply src/kernels/matern.jl]: Hahaha oops

**Comment by theogf** [comment]
Perfect I merge it then!

**Pull request merged**